Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m

```diff
@@ -161,17 +166,22 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
                 }
             }
             else if ((currentCharacterType == CharacterTypeControlCharacter)
-                     && (previousCharacterType == CharacterTypeWhitespace || previousCharacter == 0)) {
-                if (self.charactersSinceLastWhitespace == 0) {
+                     && (previousCharacterType == CharacterTypeSeparator || previousCharacter == 0)) {
+                if (self.charactersSinceLastWhitespace == 0 || previousCharacterType == CharacterTypeSeparator) {
                     // Start an EXPLICIT MENTION
+                    if (previousCharacterType == CharacterTypeSeparator
+                        && [HKWMentionsStartDetectionStateMachine.punctuationSet characterIsMember:previousCharacter]) {
+                        self.stringBuffer = [@"" mutableCopy];
+                    }
                     self.state = HKWMentionsStartDetectionStateCreatingMention;
                     [delegate beginMentionsCreationWithString:[self.stringBuffer copy]
                                                    alreadyInserted:inserted
                                              usingControlCharacter:YES
                                                   controlCharacter:c];
                 }
             }
-            else if (currentCharacterType == CharacterTypeWhitespace) {
+            else if (currentCharacterType == CharacterTypeSeparator
+                     && [HKWMentionsStartDetectionStateMachine.whitespaceSet characterIsMember:c]) {
                 // User typed a whitespace/newline. Reset the counter.
                 self.charactersSinceLastWhitespace = 0;
```

Comment:
let's not do it in this RB, but if whitespace/punctuation are now equivalent "separators" we can just have this "charactersSince" logic be a "charactersSinceLastSeparator"